internal/providers/vault/secret.go

```diff
@@ -29,34 +30,36 @@ func secretProvider(ctx context.Context, conf cfg.Configuration, secretId secret
 
 	return tasks.Return(ctx, tasks.Action("vault.read-secret").Arg("ref", secretRef),
 		func(ctx context.Context) ([]byte, error) {
-			secretPkg, secretKey, found := strings.Cut(secretRef, ":")
-			if !found {
-				return nil, fnerrors.BadInputError("invalid vault secret reference: expects secret refernece in format '<mount>/<path>:<key>'")
-			}
-
-			secretMount, secretPath, found := strings.Cut(secretPkg, "/")
-			if !found {
-				return nil, fnerrors.BadInputError("invalid vault secret package: expects secret package in format '<mount>/<path>'")
-			}
-			vaultClient, err := login(ctx, vaultConfig)
-			if err != nil {
-				return nil, err
-			}
-
-			secretResp, err := vaultClient.Secrets.KvV2Read(ctx, secretPath, vaultclient.WithMountPath(secretMount))
-			if err != nil {
-				return nil, fnerrors.InvocationError("vault", "failed to read a secret: %w", err)
-			}
-
-			if secretResp.Data.Data == nil {
-				return nil, fnerrors.InvocationError("vault", "secret response contained no data")
-			}
-
-			secret, ok := secretResp.Data.Data[secretKey].(string)
-			if !ok {
-				return nil, fnerrors.InvocationError("vault", "response data contained no expected secret %q", secretKey)
-			}
-
-			return []byte(secret), nil
+			return tryhard.CallSideEffectFree1(ctx, true, func(ctx context.Context) ([]byte, error) {
+				secretPkg, secretKey, found := strings.Cut(secretRef, ":")
+				if !found {
+					return nil, fnerrors.BadInputError("invalid vault secret reference: expects secret refernece in format '<mount>/<path>:<key>'")
+				}
+
+				secretMount, secretPath, found := strings.Cut(secretPkg, "/")
+				if !found {
+					return nil, fnerrors.BadInputError("invalid vault secret package: expects secret package in format '<mount>/<path>'")
+				}
+				vaultClient, err := login(ctx, vaultConfig)
```

Comment:
you wrap login call in this function, but not here - https://github.com/namespacelabs/foundation/pull/1354/files#diff-75ea27d66aada54407e9691a93b5e48da8ea56c29a6ae2f852da1bc38102bd9aR27
I am wondering if it makes sense to do retry inside `login` and move it here out of the retry function.